File: obstacle_avoidance/src/workspace_filter.py
#!/usr/bin/env python3

import rospy
from sensor_msgs.msg import PointCloud2
import numpy as np
from sensor_msgs import point_cloud2
import tf2_ros
import tf2_sensor_msgs
from tf2_sensor_msgs import tf2_sensor_msgs
import tf2_geometry_msgs
import pcl
import std_msgs
import time

class WorkspaceFilter:
    def __init__(self, base_boundaries,target_frame):
        rospy.init_node('workspace_filter')

        self.base_boundaries = base_boundaries
        self.target_frame = target_frame
        self.tf_buffer = tf2_ros.Buffer()
        tf_listener = tf2_ros.TransformListener(self.tf_buffer)
        
        self.cloud_sub = rospy.Subscriber('/camera/depth/color/points_sampled', PointCloud2, self.pointcloud_callback,buff_size=10000)

        # Create a publisher for the filtered point cloud
        self.filtered_cloud_pub = rospy.Publisher('/camera/depth/color/points_bounded', PointCloud2, queue_size=1)


    def pointcloud_callback(self, cloud_msg):
        # Transform point cloud to the desired frame
        #start = time.time()
        cloud_transformed_msg = self.transform_pointcloud(cloud_msg,target_frame)
        
        cloud_transformed = list(point_cloud2.read_points(cloud_transformed_msg, field_names=("x", "y", "z"), skip_nans=True))
        
        cloud_transformed = np.array(cloud_transformed)
        # Filter points outside the specified box area
        filtered_cloud = self.filter_pointcloud(cloud_transformed)
        
        # Publish the filtered point cloud
        self.publish_filtered_cloud(filtered_cloud, cloud_transformed_msg.header)
        #print(time.time() - start)
        #print('---')

    def transform_pointcloud(self, cloud_msg, target_frame):
        if not cloud_msg.header.frame_id == target_frame:
            try:
                # Look up the transformation from the source frame to the target frame
                transform = self.tf_buffer.lookup_transform(target_frame, cloud_msg.header.frame_id, rospy.Time(0), rospy.Duration(0))
            except (tf2_ros.LookupException, tf2_ros.ConnectivityException, tf2_ros.ExtrapolationException):
                rospy.logwarn("Transform not available!")
                return

            # Apply the transformation to the PointCloud
            transformed_pcl_pc = tf2_sensor_msgs.do_transform_cloud(cloud_msg, transform)
            transformed_pcl_pc.header.stamp=cloud_msg.header.stamp
        else:
            transformed_pcl_pc = cloud_msg

        return transformed_pcl_pc
    
    def filter_pointcloud(self, cloud):
        x_min, x_max, y_min, y_max, z_min, z_max = base_boundaries

        # Filter points outside the specified box area
        mask_x = np.logical_and(cloud[:, 0] >= x_min, cloud[:, 0] <= x_max)
        mask_y = np.logical_and(cloud[:, 1] >= y_min, cloud[:, 1] <= y_max)
        mask_z = np.logical_and(cloud[:, 2] >= z_min, cloud[:, 2] <= z_max)

        # Combine masks to get the final filter
        final_mask = np.logical_and(np.logical_and(mask_x, mask_y), mask_z)

        # Apply the filter to the point cloud
        filtered_cloud = cloud[final_mask]

        return filtered_cloud
    
    def publish_filtered_cloud(self, cloud,header):
        # Create a PointCloud2 message
        filtered_cloud_msg = point_cloud2.create_cloud_xyz32(header, cloud)

        # Publish the filtered point cloud
        self.filtered_cloud_pub.publish(filtered_cloud_msg)

if __name__ == '__main__':
    base_boundaries = [-0.45,0.45,-0.15,0.75,0.05,100] # x_min, x_max, y_min, y_max, z_min, z_max
    target_frame = 'base_link'
    transformer = WorkspaceFilter(base_boundaries, target_frame)

    rospy.spin()

File: obstacle_avoidance/src/workspace_filter.cpp
#include <ros/ros.h>

#include "obstacle_avoidance.h"

class WorkspaceFilter
{
public:
    WorkspaceFilter(ros::NodeHandle& nh) :nh_(nh)
    {
        // Topics
        std::string cloud_sub_topic;
        std::string cloud_pub_topic;
        std::string filter_volume_pub_topic;

        // Parameters update
        nh_.param<double>("x_min", x_min, -0.45);
        nh_.param<double>("x_max", x_max, 0.45);
        nh_.param<double>("y_min", y_min, -0.15);
        nh_.param<double>("y_max", y_max, 0.75);
        nh_.param<double>("z_min", z_min, 0.05);
        nh_.param<double>("z_max", z_max, 100.0);
        nh_.param<std::string>("cloud_sub_topic", cloud_sub_topic, "/camera/depth/color/points_sampled");
        nh_.param<std::string>("cloud_pub_topic", cloud_pub_topic, "/camera/depth/color/points_bounded");
        nh_.param<std::string>("filter_volume_pub_topic", filter_volume_pub_topic, "/filter_volume_marker");
        nh_.param<std::string>("target_frame", target_frame, "base_link");
        
        // TF tranfomrs
        tf_listener_.reset(new tf2_ros::TransformListener(tf_buffer_));

        // Ros publishers/subscribers
        cloud_sub = nh_.subscribe(cloud_sub_topic, 1, &WorkspaceFilter::pointcloudCallback, this);
        filtered_cloud_pub = nh_.advertise<sensor_msgs::PointCloud2>(cloud_pub_topic, 1);
        filter_volume_marker_pub = nh_.advertise<visualization_msgs::Marker>(filter_volume_pub_topic, 1);
        
        filter_volume_marker.header.frame_id = target_frame;
        filter_volume_marker.type = visualization_msgs::Marker::CUBE;
        filter_volume_marker.action = visualization_msgs::Marker::ADD;
        filter_volume_marker.scale = Point::VectorXYZ(abs(x_max-x_min),abs(y_max-y_min),abs(z_max-z_min));
        filter_volume_marker.lifetime = ros::Duration(0);
        filter_volume_marker.id = 0;
        filter_volume_marker.color = Color::RGBA(1,0,1,0.5);
        filter_volume_marker.pose.position.x = x_min + abs(x_max-x_min)/2;
        filter_volume_marker.pose.position.y = y_min + abs(y_max-y_min)/2;
        filter_volume_marker.pose.position.z = z_min + abs(z_max-z_min)/2;
        filter_volume_marker.pose.orientation.x = 0;
        filter_volume_marker.pose.orientation.y = 0;
        filter_volume_marker.pose.orientation.z = 0;
        filter_volume_marker.pose.orientation.w = 1;
    };
    
    void pointcloudCallback(const sensor_msgs::PointCloud2::ConstPtr &cloud_msg)
    {
        sensor_msgs::PointCloud2 cloud_transformed_msg = transformPointCloud(*cloud_msg);

        pcl::PointCloud<pcl::PointXYZ> cloud_transformed;
        pcl::fromROSMsg(cloud_transformed_msg, cloud_transformed);

        pcl::PointCloud<pcl::PointXYZ> filtered_cloud = filterPointCloud(cloud_transformed);

        sensor_msgs::PointCloud2 filtered_cloud_msg;
        pcl::toROSMsg(filtered_cloud, filtered_cloud_msg);
        filtered_cloud_msg.header = cloud_transformed_msg.header;

        filtered_cloud_pub.publish(filtered_cloud_msg);

        filter_volume_marker_pub.publish(filter_volume_marker);
    };

    sensor_msgs::PointCloud2 transformPointCloud(const sensor_msgs::PointCloud2 &cloud_msg)
    {
        if (cloud_msg.header.frame_id != target_frame)
        {
            try
            {
                geometry_msgs::TransformStamped transform = tf_buffer_.lookupTransform(target_frame, cloud_msg.header.frame_id, ros::Time(0), ros::Duration(0));
                sensor_msgs::PointCloud2 transformed_pcl_pc;
                tf2::doTransform(cloud_msg, transformed_pcl_pc, transform);
                transformed_pcl_pc.header.stamp = cloud_msg.header.stamp;
                return transformed_pcl_pc;
            }
            catch (tf2::TransformException &ex)
            {
                ROS_WARN_STREAM("Transform not available! " << ex.what());
                return cloud_msg;
            }
        }
        else
        {
            return cloud_msg;
        }
    };

    pcl::PointCloud<pcl::PointXYZ> filterPointCloud(const pcl::PointCloud<pcl::PointXYZ> &cloud)
    {
        pcl::PointCloud<pcl::PointXYZ> filtered_cloud;
        pcl::CropBox<pcl::PointXYZ> crop_filter;
        crop_filter.setInputCloud(cloud.makeShared());
        crop_filter.setMin(Eigen::Vector4f(x_min, y_min, z_min, 1.0));
        crop_filter.setMax(Eigen::Vector4f(x_max, y_max, z_max, 1.0));
        crop_filter.filter(filtered_cloud);

        return filtered_cloud;
    };

private:
    ros::NodeHandle nh_;
    double x_min; 
    double x_max; 
    double y_min; 
    double y_max; 
    double z_min; 
    double z_max;
    std::string target_frame;
    ros::Subscriber cloud_sub;
    ros::Publisher filtered_cloud_pub;
    ros::Publisher filter_volume_marker_pub;
    visualization_msgs::Marker filter_volume_marker;
    tf2_ros::Buffer tf_buffer_;
    std::shared_ptr<tf2_ros::TransformListener> tf_listener_;
};

int main(int argc, char **argv)
{
    ros::init(argc, argv,"workspace_filter");

    ros::NodeHandle nh("~");

    WorkspaceFilter workspace_filter(nh);

    ros::spin();

    return 0;
};
